### ISSUE TITLE
Conditionally defined variable is marked undefined in some cases

### ISSUE THREAD

**Issue by sharmaeklavya2** [python/mypy#1882]
This passes on mypy:

``` py
import six

if six.PY2:
    n = 42

def main():
    # type: () -> None
    if six.PY3:
        print("n is only defined on python 2.")
    else:
        print("n =", n)
        # ... some number crunching on n
```

However, if the number crunching part is very big, this is sometimes written like this:

``` py
import six

if six.PY2:
    n = 42

def main():
    # type: () -> None
    if six.PY3:
        print("n is only defined on python 2.")
        return
    print("n =", n)
    # ... some number crunching on n
```

This gives the error:

```
fail.py: note: In function "main":
fail.py:11: error: Name 'n' is not defined
```

I think both the forms are equivalent and so the second form shouldn't throw an error on mypy.

**Comment by gvanrossum** [comment]
Yeah, I agree this could be handled better. Probably the binder could keep track of this. But I can't give it much priority, sorry. If you're interested you can try to submit a PR, but be warned, the binder is one of mypy's more complex mechanisms.

**Comment by sharmaeklavya2** [comment]
It's fine if you want to deal with this later. It won't hurt me to put an `else` in my code until this is fixed. I don't think I can submit a pull request for this.

**Comment by rwbarton** [comment]
I think this error is produced during semantic analysis, at which point we don't yet have the type binder anyways. Semantic analysis could potentially do some of the same kind of flow control but it currently doesn't.

**Comment by ilevkivskyi** [comment]
From user point of view this still looks like a false positive, so marking it as such.

**Comment by hauntsaninja** [comment]
I think this error mostly makes sense.
